<a href="https://colab.research.google.com/github/anjanpa/Regression-House-Price-prediction-kaggle-/blob/master/House_Price_Regression_Hypertuning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
# Getting our tools ready
import pandas as pd
import numpy as plt
import numpy as np


In [2]:
## IMporting the preprocessed data uploaded from jupyter 
train_preprocessed=pd.read_csv("/content/drive/My Drive/house_price_preprocessed_data.csv")

In [3]:
train_preprocessed.head(5)

Id  MSSubClass  MSZoning  ...  SaleType  SaleCondition  SalePrice
0  1350          70         5  ...         9              5     122000
1  1138          50         4  ...         9              5      94000
2   748          70         5  ...         9              5     265979
3   305          75         5  ...         9              5     295000
4  1133          70         5  ...         9              5     117500

[5 rows x 81 columns]

# Evaluating results in different MODELS

In [7]:
## Splitting the data into train and test
np.random.seed(100)
from sklearn.model_selection import train_test_split
X=train_preprocessed.drop("SalePrice",axis=1)
y=train_preprocessed["SalePrice"]
X_train,X_valid,y_train,y_valid=train_test_split(X,
                                                 y,
                                                 test_size=0.2,
                                                 shuffle=True)

In [8]:
## Import the models
## IMporting the models for regression to experiment on those 
from sklearn.linear_model import LinearRegression,Lasso,Ridge
from sklearn.ensemble import RandomForestRegressor,GradientBoostingRegressor,AdaBoostRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor

In [9]:
## Making a dictionary of models
models={"LinearRegression":LinearRegression(),
       "Lasso":Lasso(),
       "Ridge":Ridge(),
       "RandomForestRegression":RandomForestRegressor(),
       "Gradient":GradientBoostingRegressor(),
       "AdaBoostRegressor":AdaBoostRegressor(),
       "SVR":SVR(),
       "DecisionTreeRegressor":DecisionTreeRegressor()}

In [12]:
## Making a function to fit and score on the models
from sklearn.metrics import mean_absolute_error
def fit_and_score(models,X_train,y_train,X_valid,y_valid):
    results={}
    for model_name,model in models.items():
        model.fit(X_train,y_train)
        results[model_name+" accuracy"]=model.score(X_valid,y_valid)
        y_preds=model.predict(X_valid)
        results[model_name+" mean_absolute_error"]=mean_absolute_error(y_valid,y_preds)

    return results
        

In [15]:
## Let's evaluate the function
%%time
result_1=fit_and_score(models,X_train,y_train,X_valid,y_valid)
result_1

CPU times: user 2.98 s, sys: 139 ms, total: 3.12 s
Wall time: 2.97 s


In [16]:
result_1

{'AdaBoostRegressor accuracy': 0.8288756579802785,
 'AdaBoostRegressor mean_absolute_error': 22086.153119072078,
 'DecisionTreeRegressor accuracy': 0.7467040846726105,
 'DecisionTreeRegressor mean_absolute_error': 24298.87671232877,
 'Gradient accuracy': 0.8819786831117113,
 'Gradient mean_absolute_error': 16262.283475624627,
 'Lasso accuracy': 0.7963561438273482,
 'Lasso mean_absolute_error': 19303.689239995423,
 'LinearRegression accuracy': 0.7961041368744275,
 'LinearRegression mean_absolute_error': 19309.667603418857,
 'RandomForestRegression accuracy': 0.8800196812918526,
 'RandomForestRegression mean_absolute_error': 17042.85609589041,
 'Ridge accuracy': 0.8270878164693442,
 'Ridge mean_absolute_error': 19069.553170312112,
 'SVR accuracy': -0.058692609181610685,
 'SVR mean_absolute_error': 53341.10468022511}

#### NOTE: 
**SVR is discarded due to low metric obtained and is very hard to improve**

## Hyperparameter tuning on different models using GridSearchCV

## 1.RandomForestRegressor

In [17]:
from sklearn.model_selection import GridSearchCV


In [20]:
## The parameter's values have been set using thorough research
np.random.seed(100)
randomforest_grid={"criterion":["mae"],
           "n_estimators":[100,150,200],
           "min_samples_split":[2,4],
           "max_features":[None,"sqrt"]}
RandomForest_grid=GridSearchCV(RandomForestRegressor(),
                               param_grid=randomforest_grid,
                               cv=5,
                               verbose=True)

In [21]:
## Create a function to check metrics of each model
def check_metrics(model):
    model.fit(X_train,y_train)
    print(model.best_params_)
    print("accuracy",model.score(X_valid,y_valid))
    print("mae",mean_absolute_error(y_valid,model.predict(X_valid)))

In [22]:
## Fitting on the data
check_metrics(RandomForest_grid)


Fitting 5 folds for each of 12 candidates, totalling 60 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  60 out of  60 | elapsed:  9.8min finished


{'criterion': 'mae', 'max_features': 'sqrt', 'min_samples_split': 2, 'n_estimators': 100}
accuracy 0.8660760465434928
mae 17015.360273972605


**NOte:It shows that mean_absolute_error has improved from 17042 to 17015 thought the accuracy has decreased.**

## 2.GradientBoostingRegressor

In [24]:
## Making a gradientboosting hyperparameters grid 
np.random.seed(100)
gradientboost_grid={"criterion":["mae"],
           "n_estimators":[100,150,200],
           "min_samples_split":[2,4],
           "max_features":[None,"sqrt"]}
Gradient_grid=GridSearchCV(RandomForestRegressor(),
                               param_grid=gradientboost_grid,
                               cv=5,
                               verbose=True)

In [25]:
check_metrics(Gradient_grid)

Fitting 5 folds for each of 12 candidates, totalling 60 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  60 out of  60 | elapsed:  9.7min finished


{'criterion': 'mae', 'max_features': 'sqrt', 'min_samples_split': 2, 'n_estimators': 100}
accuracy 0.8660760465434928
mae 17015.360273972605


In [40]:
def check_metrics_new(model):
    model.fit(X_train,y_train)
    print("accuracy",model.score(X_valid,y_valid))
    print("mae",mean_absolute_error(y_valid,model.predict(X_valid)))

In [44]:
for i in range(590,650,10):
  print(f"{i} estimators")
  check_metrics_new(GradientBoostingRegressor(n_estimators=i))
  print("\n")

590 estimators
accuracy 0.8852450493791971
mae 15633.017280666898


600 estimators
accuracy 0.8827160527610369
mae 15800.072721166362


610 estimators
accuracy 0.8864951020326327
mae 15691.116060834514


620 estimators
accuracy 0.8882523848405276
mae 15621.247033954407


630 estimators
accuracy 0.8853231644387081
mae 15701.377759707291


640 estimators
accuracy 0.885152815048356
mae 15679.425202089096




**NOTE:For gradientboosting regressor ,the model with hypertuning n_estimators is better and 260 found to be the best**

## 3.DecisionTreeRegressor

In [26]:

np.random.seed(100)
decision_tree_grid={"criterion":['mae'],
                    "min_samples_split":[2,4,6,8,10],
                    "max_features":["auto","sqrt"],

                    }
Decision_grid=GridSearchCV(DecisionTreeRegressor(),
                                param_grid=decision_tree_grid,
                                cv=5,
                                verbose=5)

In [27]:
check_metrics(Decision_grid)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[CV] criterion=mae, max_features=auto, min_samples_split=2 ...........


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  criterion=mae, max_features=auto, min_samples_split=2, score=0.628, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=2 ...........


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.2s remaining:    0.0s


[CV]  criterion=mae, max_features=auto, min_samples_split=2, score=0.693, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=2 ...........


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.4s remaining:    0.0s


[CV]  criterion=mae, max_features=auto, min_samples_split=2, score=0.801, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=2 ...........


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.7s remaining:    0.0s


[CV]  criterion=mae, max_features=auto, min_samples_split=2, score=0.524, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=2 ...........


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    0.9s remaining:    0.0s


[CV]  criterion=mae, max_features=auto, min_samples_split=2, score=0.777, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=4 ...........
[CV]  criterion=mae, max_features=auto, min_samples_split=4, score=0.620, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=4 ...........
[CV]  criterion=mae, max_features=auto, min_samples_split=4, score=0.707, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=4 ...........
[CV]  criterion=mae, max_features=auto, min_samples_split=4, score=0.692, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=4 ...........
[CV]  criterion=mae, max_features=auto, min_samples_split=4, score=0.530, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=4 ...........
[CV]  criterion=mae, max_features=auto, min_samples_split=4, score=0.742, total=   0.2s
[CV] criterion=mae, max_features=auto, min_samples_split=6 ...........
[CV]  criterion=mae, max_features=auto, min_sa

[Parallel(n_jobs=1)]: Done  50 out of  50 | elapsed:    6.0s finished


{'criterion': 'mae', 'max_features': 'auto', 'min_samples_split': 10}
accuracy 0.7392098140402124
mae 24701.532534246577


**NOTE:For decision tree regressor ,the model without hypertuning was better**

**FINAL NOTE:WITH ALL THOSE TUNING WE FOUND OUT THAT GradientBoostingRegressor IS THE BEST FOR THIS PROBLEM.**

# Train the model on full data 

In [46]:
##Splitting the data
X=train_preprocessed.drop("SalePrice",axis=1)
y=train_preprocessed["SalePrice"]

In [47]:
model=GradientBoostingRegressor(n_estimators=260).fit(X,y)

In [49]:
## Exporting the model
import pickle
pickle.dump(model,open("house_regression_model.pkl","wb"))

## Since while importing the model in pkl format in jupyter, it showed warning,we predicted on the test data here  by importing preprocessed test data**

In [50]:
## importing the test data
test=pd.read_csv("/content/drive/My Drive/test_preprocessed_house_regression.csv")
test.head()

Id  MSSubClass  MSZoning  ...  YrSold  SaleType  SaleCondition
0  1461          20         3  ...    2010         9              5
1  1462          20         4  ...    2010         9              5
2  1463          60         4  ...    2010         9              5
3  1464          60         4  ...    2010         9              5
4  1465         120         4  ...    2010         9              5

[5 rows x 80 columns]

##NOTE: TEST DATA PREPROCESSING IS DONE ON ANOTHER NOTEBOOK WHICH YOU CAN SEE IF YOU GO THROUGH THE REPOSITORY

## Predicting on the model

In [51]:
predictions=model.predict(test)
predictions

array([119158.68487262, 165505.87507234, 181130.61602463, ...,
       159005.88488545, 117619.24451931, 235075.3247618 ])

In [53]:
## Creating data frame 
submit_to_kaggle_colab=pd.DataFrame(columns=['Id','SalePrice'])
submit_to_kaggle_colab

Empty DataFrame
Columns: [Id, SalePrice]
Index: []

In [54]:
submit_to_kaggle_colab["Id"],submit_to_kaggle_colab["SalePrice"]=test["Id"],predictions
submit_to_kaggle_colab

Id      SalePrice
0     1461  119158.684873
1     1462  165505.875072
2     1463  181130.616025
3     1464  184503.432843
4     1465  183681.246909
...    ...            ...
1454  2915   72891.347321
1455  2916   75905.356421
1456  2917  159005.884885
1457  2918  117619.244519
1458  2919  235075.324762

[1459 rows x 2 columns]

In [55]:
## Exporting the data to be submitted in the kaggle
submit_to_kaggle_colab.to_csv("submit_to_kaggle_colab.csv",index=False)